### PR TITLE
Use a DB for all key stores

### PR DESCRIPTION
**Pull request by darioAnongba**
## PR description
This PR moves the business logic to the connector and uses the stores as data layers to write and read data from the underlying Vaults.
The connectors now become the list of use cases of our domains (here keys only) and orchestrate the flow between logger, vault and DB. The interface of connectors is not the same as the store interface anymore.

## Fixed Issue(s)
Part of #231 

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://github.com/consensys/quorum-key-manager/blob/master/CHANGELOG.md).

### CODE REVIEW

File: src/stores/connectors/keys/create.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+	logger.Debug("creating key")
+
+	key, err := c.store.Create(ctx, id, alg, attr)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.db.Keys().Add(ctx, key)
+	if err != nil {
+		return nil, err
```
Comment (6l20) [inline src/stores/connectors/keys/create.go:19]: Shouldn't we "UnCreate" the created key in that case ?

**Review by 6l20 — APPROVED** [review]
I left a few comments as questions and further discussions points.

File: src/stores/connectors/keys/delete.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+)
+
+func (c Connector) Delete(ctx context.Context, id string) error {
+	logger := c.logger.With("id", id)
+	logger.Debug("deleting key")
+
+	err := c.db.RunInTransaction(ctx, func(dbtx database.Database) error {
+		err := dbtx.Keys().Delete(ctx, id)
+		if err != nil {
+			return err
```
Comment (ggarri) [inline src/stores/connectors/keys/delete.go:15]: This error will be handled to the users, we should be logging here and rephasing the message to something like "failed to persist key"

File: src/stores/connectors/keys/list.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
+package keys
+
+import "context"
+
+func (c Connector) List(ctx context.Context) ([]string, error) {
+	ids := []string{}
+	keysRetrieved, err := c.db.Keys().GetAll(ctx)
```
Comment (ggarri) [inline src/stores/connectors/keys/list.go:7]: Alternatively we could read IDs directly from DB for optimization proposes

File: src/stores/store/models/testutils/faker.go [hunk 4]
```diff
@@ -1,9 +1,9 @@
 package testutils
 
 import (
+	"github.com/consensys/quorum-key-manager/src/stores/store/models"
 	"time"
 
-	"github.com/consensys/quorum-key-manager/src/stores/store/database/models"
```
Comment (ggarri) [inline src/stores/store/models/testutils/faker.go:6]: Why did you move your models one level up, as far as I see it models do not make sense out of database/, same as migrations

File: src/stores/connectors/keys/get.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+import (
+	"context"
+	"github.com/consensys/quorum-key-manager/src/stores/store/entities"
+)
+
+func (c Connector) Get(ctx context.Context, id string) (*entities.Key, error) {
+	logger := c.logger.With("id", id)
+
+	key, err := c.db.Keys().Get(ctx, id)
```
Comment (ggarri) [inline src/stores/connectors/keys/get.go:11]: is this query excluding soft_deleted items ?

File: src/stores/connectors/keys/list.go (hunk 3 above)
Comment (ggarri) [reply src/stores/connectors/keys/list.go:7]: is this query excluding soft_deleted items ?

File: src/stores/connectors/keys/update.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+	logger.Debug("updating key")
+
+	key, err := c.db.Keys().Get(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	key.Tags = attr.Tags
+
+	err = c.db.RunInTransaction(ctx, func(dbtx database.Database) error {
+		derr := c.db.Keys().Update(ctx, key)
```
Comment (ggarri) [inline src/stores/connectors/keys/update.go:20]: is go_pg automatically updating the field `updated_at` or should you take care of it?

File: src/stores/store/eth1/local/parser.go [hunk 7]
```diff
@@ -1,12 +1,12 @@
...
 
 import (
-	"github.com/consensys/quorum-key-manager/src/stores/store/database/models"
 	"github.com/consensys/quorum-key-manager/src/stores/store/entities"
+	"github.com/consensys/quorum-key-manager/src/stores/store/models"
 	"github.com/ethereum/go-ethereum/crypto"
 )
 
-func parseKey(key *entities.Key, attr *entities.Attributes) *models.ETH1Account {
+func parseKey(key *models.Key, attr *entities.Attributes) *models.ETH1Account {
```
Comment (ggarri) [inline src/stores/store/eth1/local/parser.go:9]: Trying to apply a more strict idea of how Hex Arch work, models should never live outside of our /database, our entities are the respentation of our application logic and those are the elements we should move around. Therefore I would suggest to encapsulate all this under /database aiming to prevent same issues we go in Orchestrate where we got recurrent questions about when we should use models o entities.

File: src/stores/store/keys/akv/akv.go [hunk 8]
```diff
@@ -30,11 +28,7 @@ func New(client akv.KeysClient, logger log.Logger) *Store {
 	}
 }
 
-func (s *Store) Info(context.Context) (*entities.StoreInfo, error) {
-	return nil, errors.ErrNotImplemented
-}
-
-func (s *Store) Create(ctx context.Context, id string, alg *entities.Algorithm, attr *entities.Attributes) (*entities.Key, error) {
+func (s *Store) Create(ctx context.Context, id string, alg *entities.Algorithm, attr *entities.Attributes) (*models.Key, error) {
```
Comment (ggarri) [inline src/stores/store/keys/akv/akv.go:31]: IMO stores should always work with our entities, and models are a external reprensetation of our storage which never should leak into our application to reduce the potential impact of changes over our DB model

File: src/stores/connectors/keys/create.go [hunk 9]
```diff
@@ -0,0 +1,24 @@
+package keys
```
[the remark this answers is not in the document]
Comment (darioAnongba) [reply src/stores/connectors/keys/create.go:1]: I prefer one file per use case than one big file that implements every use case and exceeds the 300 lines.

File: src/stores/connectors/keys/import.go [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+import (
+	"context"
+	"github.com/consensys/quorum-key-manager/src/stores/store/entities"
+)
+
+func (c Connector) Import(ctx context.Context, id string, privKey []byte, alg *entities.Algorithm, attr *entities.Attributes) (*entities.Key, error) {
+	logger := c.logger.With("id", id).With("algorithm", alg.Type).With("curve", alg.EllipticCurve)
+	logger.Debug("importing key")
+
+	key, err := c.store.Import(ctx, id, privKey, alg, attr)
```
[the remark this answers is not in the document]
Comment (darioAnongba) [reply src/stores/connectors/keys/import.go:12]: how would you implement this otherwise? We need a mechanism to encrypt and decrypt private keys when you want to import them? Every wallet out there (Ledger, Metamask, etc.) allows you to import a "Seed", which is your private key, the difference is that these wallets are local

File: src/stores/connectors/keys/verify.go [hunk 11]
```diff
@@ -0,0 +1,39 @@
+package keys
+
+import (
+	"context"
+	"github.com/consensys/quorum-key-manager/pkg/crypto"
+	"github.com/consensys/quorum-key-manager/pkg/errors"
+	"github.com/consensys/quorum-key-manager/src/stores/store/entities"
+)
+
+func (c Connector) Verify(_ context.Context, pubKey, data, sig []byte, algo *entities.Algorithm) error {
```
[the remark this answers is not in the document]
Comment (darioAnongba) [reply src/stores/connectors/keys/verify.go:10]: no, pubKey is a parameter of the function. You are not verifying that a signature belongs to a key in your DB, you are verifying any signature belongs to any pubKey. This function should not even be under "stores" at all.

File: src/stores/store/eth1/local/parser.go (hunk 7 above)
Comment (dolanor) [reply src/stores/store/eth1/local/parser.go:9]: It seems again we have a different naming of things.
I thought that entities were directly linked to database rows; and models more connected to the business/domain.

File: src/stores/connectors/keys/list.go (hunk 3 above)
Comment (darioAnongba) [reply src/stores/connectors/keys/list.go:7]: won't have time but I created a ticket: https://app.zenhub.com/workspaces/orchestrate-5ea70772b186e10067f57842/issues/consensys/quorum-key-manager/241

File: src/stores/store/models/testutils/faker.go (hunk 4 above)
Comment (darioAnongba) [reply src/stores/store/models/testutils/faker.go]: reverted

File: src/stores/connectors/keys/get.go (hunk 5 above)
Comment (darioAnongba) [reply src/stores/connectors/keys/get.go:12]: adding tests

File: src/stores/connectors/keys/update.go (hunk 6 above)
Comment (darioAnongba) [reply src/stores/connectors/keys/update.go]: adding tests

File: src/stores/store/eth1/local/parser.go (hunk 7 above)
Comment (darioAnongba) [reply src/stores/store/eth1/local/parser.go]: moved to using models

File: src/stores/store/keys/akv/akv.go (hunk 8 above)
Comment (darioAnongba) [reply src/stores/store/keys/akv/akv.go]: moved to using models

File: src/stores/connectors/keys/import.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (darioAnongba) [reply src/stores/connectors/keys/import.go:13]: letting this ticket to a different topic

File: src/stores/connectors/keys/create.go (hunk 1 above)
Comment (darioAnongba) [reply src/stores/connectors/keys/create.go:20]: to discuss

Comment (darioAnongba) [reply src/stores/connectors/keys/create.go:20]: won't have time to discuss this so I created a ticket: https://app.zenhub.com/workspaces/orchestrate-5ea70772b186e10067f57842/issues/consensys/quorum-key-manager/242

File: src/stores/connectors/keys/get.go (hunk 5 above)
Comment (darioAnongba) [reply src/stores/connectors/keys/get.go:12]: tests added

File: src/stores/connectors/keys/update.go (hunk 6 above)
Comment (darioAnongba) [reply src/stores/connectors/keys/update.go]: tests added
